src/messages/MOSDRepOpReply.h

```diff
@@ -46,20 +46,50 @@ class MOSDRepOpReply : public Message {
   // piggybacked osd state
   eversion_t last_complete_ondisk;
 
+  bufferlist::iterator p;
+  // Decoding flags. Decoding is only needed for messages catched by pipe reader.
+  bool partial_decode_needed;
+  bool final_decode_needed;
 
   virtual void decode_payload() {
-    bufferlist::iterator p = payload.begin();
-    ::decode(map_epoch, p);
-    ::decode(reqid, p);
-    ::decode(pgid, p);
+    p = payload.begin();
+    if (header.version < 2) {
+      ::decode(map_epoch, p);
+      ::decode(reqid, p);
+      ::decode(pgid, p);
+
+      ::decode(ack_type, p);
+      ::decode(result, p);
+      ::decode(last_complete_ondisk, p);
+
+      ::decode(from, p);
+    } else {
+      // new, v2 decode, splitted to partial and final
+      ::decode(map_epoch, p);
+      ::decode(reqid, p);
+      ::decode(pgid, p);
+    }
+    partial_decode_needed = false;
+  }
 
+  void finish_decode() {
+    assert(!partial_decode_needed); // partial decoding required
+    if (!final_decode_needed)
+      return; // Message is already final decoded
+    assert(header.version >= 2);
     ::decode(ack_type, p);
     ::decode(result, p);
     ::decode(last_complete_ondisk, p);
 
     ::decode(from, p);
+    final_decode_needed = false;
```

Comment:
Yes, I think since we have split the MOSDRepOp msg into two stages, split the MOSDRepOpReply require minimal effort.